tmt/frameworks/shell.py

```diff
@@ -23,30 +21,118 @@ def get_test_command(
+
     @classmethod
     def extract_results(
             cls,
             invocation: 'TestInvocation',
+            results: list[tmt.result.Result],
             logger: tmt.log.Logger) -> list[tmt.result.Result]:
-        """ Check result of a shell test """
+        """
+
+        Check result of a shell test.
+
+        If there are no extra results (e.g. extracted from the tmt-report-results.yaml), continue
+        normally - set the parent result outcome according to test exit status.
+
+        Otherwise, process given results, reduce their outcomes into a single one and set these
+        results as tmt subresults.
+
+        :param invocation: test invocation to which the results belong to.
+        :param results: results to reduce and save as tmt subresults.
+        :returns: list of results.
+        """
         assert invocation.return_code is not None
         note = None
 
-        try:
-            # Process the exit code and prepare the log path
-            result = {0: ResultOutcome.PASS, 1: ResultOutcome.FAIL}[invocation.return_code]
-        except KeyError:
-            result = ResultOutcome.ERROR
-            # Add note about the exceeded duration
-            if invocation.return_code == tmt.utils.ProcessExitCodes.TIMEOUT:
-                note = 'timeout'
-                invocation.phase.timeout_hint(invocation)
-
-            elif tmt.utils.ProcessExitCodes.is_pidfile(invocation.return_code):
-                note = 'pidfile locking'
+        # If no extra results were passed (e.g. `tmt-report-result` was not called during the
+        # test), just process the exit code of a shell test and return the result.
+        if not results:
+            try:
+                # Process the exit code and prepare the log path
+                result = {0: ResultOutcome.PASS, 1: ResultOutcome.FAIL}[invocation.return_code]
+            except KeyError:
+                result = ResultOutcome.ERROR
+                # Add note about the exceeded duration
+                if invocation.return_code == tmt.utils.ProcessExitCodes.TIMEOUT:
+                    note = 'timeout'
+                    invocation.phase.timeout_hint(invocation)
 
-        return [tmt.Result.from_test_invocation(
-            invocation=invocation,
-            result=result,
-            log=[invocation.relative_path / tmt.steps.execute.TEST_OUTPUT_FILENAME],
-            note=note)]
+                elif tmt.utils.ProcessExitCodes.is_pidfile(invocation.return_code):
+                    note = 'pidfile locking'
+
+            return [tmt.Result.from_test_invocation(
+                invocation=invocation,
+                result=result,
+                log=[invocation.relative_path / tmt.steps.execute.TEST_OUTPUT_FILENAME],
+                note=note)]
+
+        # Handle the `tmt-report-result` command results as a single test with assigned tmt
+        # subresults.
+        return cls._process_results_reduce(invocation, results)
```

Comment:
If you switch the condition above and switch the order, `if results:`, you can `return cls._process_results_reduce(invocation, results)` right from that branch above, and the whole block under `if not results:` could be indented one level to the left, which could make a slightly better diff - the code probably did not change that much, it's the extra indentation that inflates it.

Reply:
Yes, of course! It’s strange that I didn’t notice this when I wrote the code. Thanks for pointing it out :).